plutarch-orphanage/src/PlutusLedgerApi/V2/Orphans.hs

```diff
@@ -0,0 +1,1271 @@
+using 'PlutusLedgerApi.QuickCheck.Utils.NonAdaTokenName' instead.
+
+@since 1.0.0
+-}
+instance Arbitrary PLA.TokenName where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary =
+    fmap (PLA.TokenName . PlutusTx.toBuiltin @ByteString . BS.pack) . sized $ \size -> do
+      -- We want the length to be size-dependent
+      len <- scale (const size) . chooseInt $ (0, 32)
+      -- But the bytes themselves should not be: the whole ASCII printable range
+      -- should be available always
+      vectorOf len . chooseBoundedIntegral $ (33, 126)
+  {-# INLINEABLE shrink #-}
+  shrink tn =
+    PLA.TokenName . PlutusTx.toBuiltin @ByteString <$> do
+      let asList = BS.unpack . PlutusTx.fromBuiltin @PlutusTx.BuiltinByteString . coerce $ tn
+      bs <- BS.pack <$> shrink asList
+      guard (BS.all (\w8 -> w8 >= 33 && w8 <= 126) bs)
+      pure bs
+
+-- | @since 1.0.0
+deriving via PlutusTx.BuiltinByteString instance CoArbitrary PLA.TokenName
+
+-- | @since 1.0.0
+instance Function PLA.TokenName where
+  {-# INLINEABLE function #-}
+  function = functionMap coerce PLA.TokenName
+
+{- | This generates well-defined maps: specifically, there are no duplicate
+keys. To ensure that this is preserved, we do not shrink keys: we only drop
+whole entries, or shrink values associated with keys.
+
+In order to make this instance even moderately efficient, we require an 'Ord'
+constraint on keys. In practice, this isn't a significant limitation, as
+basically all Plutus types have such an instance.
+
+@since 1.0.0
+-}
+instance (Arbitrary k, Ord k) => Arbitrary1 (AssocMap.Map k) where
+  {-# INLINEABLE liftArbitrary #-}
+  liftArbitrary genVal =
+    AssocMap.unsafeFromList <$> do
+      -- First, generate a Set of keys to ensure no duplication
+      keyList <- Set.toList <$> arbitrary
+      -- Then generate a value for each
+      traverse (\key -> (key,) <$> genVal) keyList
+  {-# INLINEABLE liftShrink #-}
+  liftShrink shrinkVal aMap =
+    AssocMap.unsafeFromList <$> do
+      let asList = AssocMap.toList aMap
+      liftShrink (\(key, val) -> (key,) <$> shrinkVal val) asList
+
+-- | @since 1.0.0
+instance (Arbitrary k, Arbitrary v, Ord k) => Arbitrary (AssocMap.Map k v) where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = liftArbitrary arbitrary
+  {-# INLINEABLE shrink #-}
+  shrink = liftShrink shrink
+
+-- | @since 1.0.0
+instance (CoArbitrary k, CoArbitrary v) => CoArbitrary (AssocMap.Map k v) where
+  {-# INLINEABLE coarbitrary #-}
+  coarbitrary = coarbitrary . AssocMap.toList
+
+-- | @since 1.0.0
+instance (Function k, Function v) => Function (AssocMap.Map k v) where
+  {-# INLINEABLE function #-}
+  function = functionMap AssocMap.toList AssocMap.unsafeFromList
+
+-- TODO: We should probably treat Ada values differently?
+
+{- | This is the most general possible instance for 'PLA.Value'. In particular,
+this can have zero values, and does not treat the Ada symbol or token name
+specially.
+
+@since 1.0.0
+-}
```

Comment:
This is where it gets tricky. UTxOValue must always have positive ADA, whereas MintValue must always have a 0 ADA.

So depends in which context you're generating for this might not be correct.

Reply:
Given this, it suggests we want two `newtype` wrappers: `UTxOValue` and `MintValue`, both of which can be 'unwrapped' to make a `Value`. I'm a bit hesitant to define 'open' modifiers, but in this case, I think it's OK.